README.md
# Terraform Provider for Philips Hue Bridge

## Quick Starts

- [Using the provider](https://registry.terraform.io/providers/AliAllomani/hue/latest)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2021-07-04

Initial Release


hue/provider_test.go
package hue

import (
	"context"
	"os"
	"sync"
	"testing"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
)

const (
	// Provider name for single configuration testing
	ProviderName = "hue"
)

// testAccProviderFactories is a static map containing only the main provider instance
//
// Use other testAccProviderFactories functions, such as testAccProviderFactoriesAlternate,
// for tests requiring special provider configurations.
var testAccProviderFactories map[string]func() (*schema.Provider, error)

// testAccProvider is the "main" provider instance
//
// This Provider can be used in testing code for API calls without requiring
// the use of saving and referencing specific ProviderFactories instances.
//
// testAccPreCheck(t) must be called before using this provider instance.
var testAccProvider *schema.Provider

// testAccProviderConfigure ensures testAccProvider is only configured once
//
// The testAccPreCheck(t) function is invoked for every test and this prevents
// extraneous reconfiguration to the same values each time. However, this does
// not prevent reconfiguration that may happen should the address of
// testAccProvider be errantly reused in ProviderFactories.
var testAccProviderConfigure sync.Once

func init() {
	testAccProvider = Provider()

	// Always allocate a new provider instance each invocation, otherwise gRPC
	// ProviderConfigure() can overwrite configuration during concurrent testing.
	testAccProviderFactories = map[string]func() (*schema.Provider, error){
		ProviderName: func() (*schema.Provider, error) { return Provider(), nil }, //nolint:unparam
	}
}

func testAccPreCheck(t *testing.T) {
	// Since we are outside the scope of the Terraform configuration we must
	// call Configure() to properly initialize the provider configuration.
	testAccProviderConfigure.Do(func() {
		if os.Getenv("HUE_HOST") == "" && os.Getenv("HUE_USER") == "" {
			t.Fatal("HUE_HOST and HUE_USER must be set for acceptance tests")
		}

		err := testAccProvider.Configure(context.Background(), terraform.NewResourceConfigRaw(nil))
		if err != nil {
			t.Fatal(err)
		}
	})
}


hue/resource_hue_light_test.go
package hue

import (
	"fmt"
	"testing"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
)

func TestAccResourceHueLight_basic(t *testing.T) {

	resourceName := "hue_light.test"
	resource.ParallelTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: testAccProviderFactories,
		CheckDestroy:      testAccResourceCheckHueLightDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccResourceHueLightConfig("65535", "240"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr(resourceName, "state.0.on", "true"),
					resource.TestCheckResourceAttr(resourceName, "light_index", "1"),
					resource.TestCheckResourceAttr(resourceName, "state.0.brightness", "240"),
				),
			},
			{
				Config: testAccResourceHueLightConfig("46920", "100"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr(resourceName, "state.0.on", "true"),
					resource.TestCheckResourceAttr(resourceName, "light_index", "1"),
					resource.TestCheckResourceAttr(resourceName, "state.0.brightness", "100"),
				),
			},
			{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
		},
	})
}

func testAccResourceCheckHueLightDestroy(s *terraform.State) error {
	// ignore check-destroy as the lights will not be removed from the bridge
	return nil
}

func testAccResourceHueLightConfig(hue string, bri string) string {
	return fmt.Sprintf(`

data "hue_lights" "test" {}

resource "hue_light" "test" {
  unique_id = data.hue_lights.test.lights.0.unique_id

  state {
    on         = true
    hue        = "%s"
    brightness = "%s"
  }
}
`, hue, bri)
}


hue/datasource_hue_light_test.go
package hue

import (
	"fmt"
	"testing"

	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
)

func TestAccDataSourceHueLight_basic(t *testing.T) {

	resourceName := "data.hue_light.test"
	resource.ParallelTest(t, resource.TestCase{
		PreCheck:          func() { testAccPreCheck(t) },
		ProviderFactories: testAccProviderFactories,
		Steps: []resource.TestStep{
			{
				Config: testAccDataSourceHueLightConfig("65535", "240"),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttr(resourceName, "state.0.on", "true"),
					resource.TestCheckResourceAttr(resourceName, "light_index", "1"),
				),
			},
		},
	})
}

func testAccDataSourceHueLightConfig(hue string, bri string) string {
	return fmt.Sprintf(`

	%s

data "hue_light" "test" {
  unique_id = hue_light.test.id
}
`, testAccResourceHueLightConfig(hue, bri))
}


hue/provider.go
package hue

import (
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func Provider() *schema.Provider {
	p := &schema.Provider{
		Schema: map[string]*schema.Schema{
			"host": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Domain name/IP of the Hue Bridge",
				DefaultFunc: schema.EnvDefaultFunc("HUE_HOST", nil),
			},
			"username": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Username of Hue Bridge",
				DefaultFunc: schema.EnvDefaultFunc("HUE_USER", nil),
			},
		},
		DataSourcesMap: map[string]*schema.Resource{
			"hue_lights": dataSourceHueLights(),
			"hue_light":  dataSourceHueLight(),
		},
		ResourcesMap: map[string]*schema.Resource{
			"hue_light": resourceHueLight(),
		},
	}
	p.ConfigureFunc = func(d *schema.ResourceData) (interface{}, error) {
		terraformVersion := p.TerraformVersion
		if terraformVersion == "" {
			// Terraform 0.12 introduced this field to the protocol
			// We can therefore assume that if it's missing it's 0.10 or 0.11
			terraformVersion = "0.11+compatible"
		}
		return providerConfigure(d, terraformVersion)
	}
	return p
}

func providerConfigure(d *schema.ResourceData, terraformVersion string) (interface{}, error) {
	config := Config{
		Host:             d.Get("host").(string),
		Username:         d.Get("username").(string),
		terraformVersion: terraformVersion,
	}

	cfg, err := config.Client()
	if err != nil {
		return cfg, err
	}

	return cfg, err
}


hue/hue_light.go
package hue

import (
	"fmt"
	"log"
	"sort"

	"github.com/amimof/huego"
)

func flattenLights(input []huego.Light) []map[string]interface{} {

	var lights []map[string]interface{}
	var state []map[string]interface{}

	for _, light := range input {

		state = make([]map[string]interface{}, 0, 1)
		state = append(state, flattenLightState(light.State))

		lights = append(lights, map[string]interface{}{
			"light_index": light.ID,
			"name":        light.Name,
			"model_id":    light.ModelID,
			"product_id":  light.ProductID,
			"unique_id":   light.UniqueID,
			"sw_version":  light.SwVersion,
			"state":       state,
		})
	}

	sort.Slice(lights, func(i, j int) bool {
		return lights[i]["light_index"].(int) < lights[j]["light_index"].(int)
	})

	return lights
}

func flattenLightState(state *huego.State) map[string]interface{} {
	flattenstate := map[string]interface{}{
		"hue":        state.Hue,
		"on":         state.On,
		"color_mode": state.ColorMode,
		"scene":      state.Scene,
		"brightness": state.Bri,
		"saturation": state.Sat,
	}
	return flattenstate
}

func lightIndexFromUniqueId(client *huego.Bridge, uniqueID string) (int, error) {

	lights, err := client.GetLights()
	if err != nil {
		return 0, fmt.Errorf("Error retrieving lights: %v", err)
	}

	for _, light := range lights {
		if light.UniqueID == uniqueID {
			log.Printf("[INFO] Found Light Index %d for Light Unique ID: %s", light.ID, light.UniqueID)
			return light.ID, nil
		}
	}

	return 0, fmt.Errorf("Cloud not find a light with Unique ID : %s", uniqueID)

}


scripts/run-tflint.sh
#!/usr/bin/env bash

function tfProviderLint {
  echo "==> Checking source code against terraform provider linters..."
	tfproviderlint \
        -AT001\
        -AT001.ignored-filename-suffixes _data_source_test.go\
        -AT005 -AT006 -AT007\
        -R001 -R002 -R003 -R004 -R006\
        -S001 -S002 -S003 -S004 -S005 -S006 -S007 -S008 -S009 -S010 -S011 -S012 -S013 -S014 -S015 -S016 -S017 -S018 -S019 -S020\
        -S021 -S022 -S023 -S024 -S025 -S026 -S027 -S028 -S029 -S030 -S031 -S032 -S033\
        ./hue/...
}

function tfAccTestsLint {
  echo "==> Checking acceptance test terraform blocks are formatted..."

files=$(find ./hue -type f -name "*_test.go")
error=false

for f in $files; do
  terrafmt diff -c -q -f "$f" || error=true
done

if ${error}; then
  echo "------------------------------------------------"
  echo ""
  echo "The preceding files contain terraform blocks that are not correctly formatted or contain errors."
  echo "You can fix this by running make tools and then terrafmt on them."
  echo ""
  echo "to easily fix all terraform blocks:"
  echo "$ make tffmtfix"
  echo ""
  echo "format only acceptance test config blocks:"
  echo "$ find hue | egrep \"_test.go\" | sort | while read f; do terrafmt fmt -f \$f; done"
  echo ""
  echo "format a single test file:"
  echo "$ terrafmt fmt -f ./hue/resource_test.go"
  echo ""
  exit 1
fi

exit 0

}

function tfDocsLint {
  echo "==> Checking docs terraform blocks are formatted..."

files=$(find ./docs -type f -name "*.md")
error=false

for f in $files; do
  terrafmt diff -c -q -f "$f" || error=true
done

if ${error}; then
  echo "------------------------------------------------"
  echo ""
  echo "The preceding files contain terraform blocks that are not correctly formatted or contain errors."
  echo "You can fix this by running make tools and then terrafmt on them."
  echo ""
  echo "to easily fix all terraform blocks:"
  echo "$ make tffmtfix"
  echo ""
  echo "format only docs config blocks:"
  echo "$ find docs | egrep \".md\" | sort | while read f; do terrafmt fmt -f \$f; done"
  echo ""
  echo "format a single test file:"
  echo "$ terrafmt fmt -f ./docs/resources/resource.md"
  echo ""
  exit 1
fi

exit 0

}

function main {
  tfProviderLint
  tfAccTestsLint
  tfDocsLint

}

main


hue/resource_hue_light.go
package hue

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/amimof/huego"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func resourceHueLight() *schema.Resource {
	return &schema.Resource{
		Create: resourceHueLightCreateUpdate,
		Read:   resourceHueLightRead,
		Update: resourceHueLightCreateUpdate,
		Delete: resourceHueLightDelete,
		Importer: &schema.ResourceImporter{
			State: schema.ImportStatePassthrough,
		},
		Schema: map[string]*schema.Schema{
			"unique_id": {
				Type:        schema.TypeString,
				Required:    true,
				ForceNew:    true,
				Description: "Unique id of the device. The MAC address of the device with a unique endpoint id in the form: AA:BB:CC:DD:EE:FF:00:11-XX",
			},
			"light_index": {
				Type:        schema.TypeInt,
				Computed:    true,
				Description: "Hue light index (ID)",
			},
			"name": {
				Type:        schema.TypeString,
				Optional:    true,
				Computed:    true,
				Description: "A unique, editable name given to the light.",
			},
			"model_id": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "The hardware model of the light.",
			},
			"product_id": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "The Product ID of the Light",
			},
			"sw_version": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "An identifier for the software version running on the light.",
			},
			"state": {
				Type:        schema.TypeList,
				MaxItems:    1,
				Computed:    true,
				Optional:    true,
				Description: "The current state of the Light",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"hue": {
							Type:        schema.TypeInt,
							Optional:    true,
							Computed:    true,
							Description: "The hue value to set light to.The hue value is a wrapping value between 0 and 65535. Both 0 and 65535 are red, 25500 is green and 46920 is blue.",
						},
						"color_mode": {
							Type:        schema.TypeString,
							Optional:    true,
							Computed:    true,
							Description: "Indicates the color mode in which the light is working, this is the last command type it received. Values are “hs” for Hue and Saturation, “xy” for XY and “ct” for Color Temperature. This parameter is only present when the light supports at least one of the values",
						},
						"on": {
							Type:        schema.TypeBool,
							Optional:    true,
							Computed:    true,
							Description: "On/Off state of the light. On=true, Off=false",
						},
						"scene": {
							Type:        schema.TypeString,
							Optional:    true,
							Computed:    true,
							Description: "The Scene name",
						},
						"brightness": {
							Type:        schema.TypeInt,
							Optional:    true,
							Computed:    true,
							Description: "The brightness value to set the light to. Brightness is a scale from 1 (the minimum the light is capable of) to 254 (the maximum).",
						},
						"saturation": {
							Type:        schema.TypeInt,
							Optional:    true,
							Computed:    true,
							Description: "Saturation of the light. 254 is the most saturated (colored) and 0 is the least saturated (white).",
						},
					},
				},
			},
		},
	}
}

func resourceHueLightCreateUpdate(d *schema.ResourceData, meta interface{}) error {

	client := meta.(*huego.Bridge)
	uniqueId := d.Get("unique_id").(string)

	lightIndex, err := lightIndexFromUniqueId(client, uniqueId)
	if err != nil {
		return fmt.Errorf("Error create/update light : %v", err)
	}

	d.SetId(uniqueId)
	d.Set("light_index", lightIndex)

	updateConfig := huego.Light{}

	if name, ok := d.GetOk("name"); ok {
		updateConfig.Name = name.(string)
	}

	updatedLight, err := client.UpdateLight(lightIndex, updateConfig)
	if err != nil {
		return fmt.Errorf("Could not Create/Update Light with Index %d : %v", lightIndex, err)
	}

	updatedLightJson, err := json.Marshal(updatedLight)
	if err != nil {
		return fmt.Errorf("Could not Marshal Light with Index %d : %v", lightIndex, err)
	}

	log.Printf("[DEBUG] Updated light with Index %d : %s", lightIndex, updatedLightJson)

	if d.HasChange("state") {
		if err := resourceHueLightStateCreateUpdate(client, d); err != nil {
			return err
		}
	}

	return resourceHueLightRead(d, meta)
}

func resourceHueLightRead(d *schema.ResourceData, meta interface{}) error {

	client := meta.(*huego.Bridge)
	uniqueId := d.Id()

	light_index, err := lightIndexFromUniqueId(client, uniqueId)
	if err != nil {
		return fmt.Errorf("Error parsing light Index from ID : %s : %v", uniqueId, err)
	}

	light, err := client.GetLight(light_index)
	if err != nil {
		return fmt.Errorf("Error retrieving light with Index %d : %v", light_index, err)
	}

	state := make([]map[string]interface{}, 0, 1)
	state = append(state, flattenLightState(light.State))

	d.Set("light_index", light.ID)
	d.Set("name", light.Name)
	d.Set("unique_id", light.UniqueID)
	d.Set("model_id", light.ModelID)
	d.Set("product_id", light.ProductID)
	d.Set("sw_version", light.SwVersion)
	d.Set("state", state)

	return nil

}

func resourceHueLightDelete(d *schema.ResourceData, meta interface{}) error {
	// We will not delete the light from the bridge, just from the state file
	return nil
}

func resourceHueLightStateCreateUpdate(client *huego.Bridge, d *schema.ResourceData) error {

	light_index := d.Get("light_index").(int)

	state := d.Get("state").([]interface{})
	updateState := huego.State{}

	var currentState map[string]interface{}
	if state[0] != nil {
		currentState = state[0].(map[string]interface{})
	} else {
		currentState = make(map[string]interface{})
	}

	if v, ok := currentState["hue"]; ok {
		updateState.Hue = uint16(v.(int))
	}

	if v, ok := currentState["on"]; ok {
		updateState.On = v.(bool)
	}

	if v, ok := currentState["scene"]; ok {
		updateState.Scene = v.(string)
	}

	if v, ok := currentState["color_mode"]; ok {
		updateState.ColorMode = v.(string)
	}

	if v, ok := currentState["brightness"]; ok {
		updateState.Bri = uint8(v.(int))
	}

	if v, ok := currentState["saturation"]; ok {
		updateState.Sat = uint8(v.(int))
	}

	updatedLightState, err := client.SetLightState(light_index, updateState)
	if err != nil {
		return fmt.Errorf("Could not Create/Update Light State with Index %d : %v", light_index, err)
	}

	updatedLightStateJson, err := json.Marshal(updatedLightState)
	if err != nil {
		return fmt.Errorf("Could not Marshal Light State with Index %d : %v", light_index, err)
	}

	log.Printf("[DEBUG] Updated light state with Index %d : %s", light_index, updatedLightStateJson)

	return nil
}


hue/datasource_hue_light.go
package hue

import (
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func dataSourceHueLight() *schema.Resource {
	return &schema.Resource{
		Read: dataSourceHueLightRead,
		Schema: map[string]*schema.Schema{
			"unique_id": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Unique id of the device. The MAC address of the device with a unique endpoint id in the form: AA:BB:CC:DD:EE:FF:00:11-XX",
			},

			"name": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "A unique, editable name given to the light.",
			},
			"light_index": {
				Type:        schema.TypeInt,
				Computed:    true,
				Description: "Hue light index (ID)",
			},
			"model_id": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "The hardware model of the light.",
			},
			"product_id": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "The Product ID of the Light",
			},
			"sw_version": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "An identifier for the software version running on the light.",
			},
			"state": {
				Type:        schema.TypeList,
				Computed:    true,
				Description: "The current state of the Light",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"hue": {
							Type:        schema.TypeInt,
							Optional:    true,
							Computed:    true,
							Description: "The hue value to set light to.The hue value is a wrapping value between 0 and 65535. Both 0 and 65535 are red, 25500 is green and 46920 is blue.",
						},
						"color_mode": {
							Type:        schema.TypeString,
							Optional:    true,
							Computed:    true,
							Description: "Indicates the color mode in which the light is working, this is the last command type it received. Values are “hs” for Hue and Saturation, “xy” for XY and “ct” for Color Temperature. This parameter is only present when the light supports at least one of the values",
						},
						"on": {
							Type:        schema.TypeBool,
							Optional:    true,
							Computed:    true,
							Description: "On/Off state of the light. On=true, Off=false",
						},
						"scene": {
							Type:        schema.TypeString,
							Optional:    true,
							Computed:    true,
							Description: "The Scene name",
						},
						"brightness": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "The brightness value to set the light to. Brightness is a scale from 1 (the minimum the light is capable of) to 254 (the maximum).",
						},
						"saturation": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "Saturation of the light. 254 is the most saturated (colored) and 0 is the least saturated (white).",
						},
					},
				},
			},
		},
	}
}

func dataSourceHueLightRead(d *schema.ResourceData, meta interface{}) error {

	uniqueId := d.Get("unique_id").(string)
	d.SetId(uniqueId)

	return resourceHueLightRead(d, meta)
}


hue/datasource_hue_lights.go
package hue

import (
	"fmt"
	"log"

	"github.com/amimof/huego"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
)

func dataSourceHueLights() *schema.Resource {
	return &schema.Resource{
		Read: dataSourceHueLightsRead,
		Schema: map[string]*schema.Schema{
			"lights": {
				Type:     schema.TypeList,
				Computed: true,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"unique_id": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "Unique id of the device. The MAC address of the device with a unique endpoint id in the form: AA:BB:CC:DD:EE:FF:00:11-XX",
						},
						"light_index": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "Hue light index (ID)",
						},
						"name": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "A unique, editable name given to the light.",
						},
						"model_id": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "The hardware model of the light.",
						},
						"product_id": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "The Product ID of the Light",
						},
						"sw_version": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "An identifier for the software version running on the light.",
						},
						"state": {
							Type:        schema.TypeList,
							Computed:    true,
							Description: "The current state of the Light",
							Elem: &schema.Resource{
								Schema: map[string]*schema.Schema{
									"hue": {
										Type:        schema.TypeInt,
										Optional:    true,
										Computed:    true,
										Description: "The hue value to set light to.The hue value is a wrapping value between 0 and 65535. Both 0 and 65535 are red, 25500 is green and 46920 is blue.",
									},
									"color_mode": {
										Type:        schema.TypeString,
										Optional:    true,
										Computed:    true,
										Description: "Indicates the color mode in which the light is working, this is the last command type it received. Values are “hs” for Hue and Saturation, “xy” for XY and “ct” for Color Temperature. This parameter is only present when the light supports at least one of the values",
									},
									"on": {
										Type:        schema.TypeBool,
										Optional:    true,
										Computed:    true,
										Description: "On/Off state of the light. On=true, Off=false",
									},
									"scene": {
										Type:        schema.TypeString,
										Optional:    true,
										Computed:    true,
										Description: "The Scene name",
									},
									"brightness": {
										Type:        schema.TypeInt,
										Computed:    true,
										Description: "The brightness value to set the light to. Brightness is a scale from 1 (the minimum the light is capable of) to 254 (the maximum).",
									},
									"saturation": {
										Type:        schema.TypeInt,
										Computed:    true,
										Description: "Saturation of the light. 254 is the most saturated (colored) and 0 is the least saturated (white).",
									},
								},
							},
						},
					},
				},
			},
		},
	}
}

func dataSourceHueLightsRead(d *schema.ResourceData, meta interface{}) error {

	client := meta.(*huego.Bridge)
	d.SetId("lights")

	lights, err := client.GetLights()
	if err != nil {
		return fmt.Errorf("Error retrieving lights: %v", err)
	}

	fLights := flattenLights(lights)

	log.Printf("[INFO] found %d lights.", len(lights))

	d.Set("lights", fLights)

	return nil
}
